Makefile

```diff
@@ -285,16 +285,23 @@ $(foreach IMAGE,$(IMAGES),clean-$(IMAGE)-image): clean-%-image:
 
 #####################################################################################################################
 update-chart-deps: build/toolchain/bin/helm$(EXE_EXTENSION)
-	(cd $(REPOSITORY_ROOT)/install/helm/open-match; $(HELM) repo add incubator https://kubernetes-charts-incubator.storage.googleapis.com; $(HELM) dependency update)
+	$(HELM) repo add incubator https://kubernetes-charts-incubator.storage.googleapis.com
+	(cd $(REPOSITORY_ROOT)/install/helm/open-match; $(HELM) dependency update)
+	(cd $(REPOSITORY_ROOT)/install/helm/open-match/subcharts/open-match-telemetry; $(HELM) dependency update)
 
 lint-chart: build/toolchain/bin/helm$(EXE_EXTENSION) build/toolchain/bin/ct$(EXE_EXTENSION)
 	(cd $(REPOSITORY_ROOT)/install/helm; $(HELM) lint $(OPEN_MATCH_HELM_NAME))
 	$(CHART_TESTING) lint --all --chart-yaml-schema $(TOOLCHAIN_BIN)/etc/chart_schema.yaml --lint-conf $(TOOLCHAIN_BIN)/etc/lintconf.yaml --chart-dirs $(REPOSITORY_ROOT)/install/helm/
 	$(CHART_TESTING) lint-and-install --all --chart-yaml-schema $(TOOLCHAIN_BIN)/etc/chart_schema.yaml --lint-conf $(TOOLCHAIN_BIN)/etc/lintconf.yaml --chart-dirs $(REPOSITORY_ROOT)/install/helm/
 
-build/chart/open-match-$(BASE_VERSION).tgz: build/toolchain/bin/helm$(EXE_EXTENSION) lint-chart
+patch-chart-image-values:
+	$(SED_REPLACE) 's|    registry: .*|    registry: $(REGISTRY)|g' $(REPOSITORY_ROOT)/install/helm/open-match/values.yaml
```

Comment:
values.yaml is commited, so running any command which calls into this would cause unwanted churn on these file, right?

Why are you replacing the current helm_image_tags?

Reply:
Because right now, the generated chart contains *floating* image tags (0.0.0-dev), and I don't think `imagePullPolicy: Always` is set here.

So installing this chart will mean your nodes are running whatever version of the image happened to be built from master at the time of first pull *for that node*. Over time, this means you'll end up with a mix of different versions of the same image, which seems a huge risk.

Reply:
Patching the registry is needed because the auto-build images are not pushed to `gcr.io/$(OPEN_MATCH_PUBLIC_IMAGES_PROJECT_ID)` (which is what's in the committed values.yaml), but to `gcr.io/$(OPEN_MATCH_BUILD_PROJECT_ID)`. 

The idea here is to test the same thing we ship. Before this change, the behaviour was to generate templates from the Helm chart, and patch _those generated files_ for CI, leaving the Helm chart untested and invalid out-of-the-box.

Another option would be to make the helm calls for CI, _and the users of the 0.0.0-dev chart_, pass `--set global.image.registry=gcr.io/$(OPEN_MATCH_BUILD_PROJECT_ID)` to Helm on each invocation (or when using it as a subchart, hard-coding that in the parent chart's values.yaml).

Reply:
Anyway, the change made by this target should not be committed, since it's patching in build-specific values.

Reply:
Right, I don't want commands making changes to files which shouldn't be committed (exception: anything in .gitignore, obviously).  At best it'd be annoying to revert, especially when there's other changes to the same file, at worst changes accidentally make there way into being committed.

Why is simply setting the variables when building the charts not the right course of action here?

Sorry for the delay reviewing this - I wasn't involved in writing the charts nor the build system so I need time to become familiar and figure out the best solution here.

Reply:
You can't set variables except `.Chart.Version` and `.Chart.AppVersion` when calling `helm package`. The best you can do is change the values.yaml before calling `helm package`. That's why I suggested as a TODO to move the image tag be `.Chart.AppVersion`, because we _can_ set that at package-build time, to remove 50% of this editing.

To get rid of the other 50% of the editing, we'd need to instead have a macro reacting to `.Chart.AppVersion` or `.Chart.Version`, and recognising an auto-build, to replace the existing hard-coded default repo with _two_ hard-coded default repos, selected by, e.g. the presence of `-dev` in `.Chart.Version`.

However, both of the above changes are complicated by the current muliti-chart setup, as the sub-charts _always_ have version `0.0.0-dev`, even in release builds, because we don't use `helm package` for them, but source them inline, which means there's no opportunity to set `.Chart.Version` and `.Chart.AppVersion` at package time.

We could, as a separate change, modify the build system to do `helm package` for the subcharts rather than `helm dep update` on the parent charts, perhaps. But per #1244, the whole subcharting setup in Open Match is pretty messy, and I'd rather spent the time unwinding that, if I was going to spend time on it. (And any Open Match time I spend has higher priorities, like the outstanding documentation updates for #1235).

Perhaps with some clever scripting, the values.yaml changes could be reverted as part of the Make execution? Although obviously that won't fly if the process fails and the user just runs it again, because then on next run, it'd be changing _from_ the modified files anyway, and change them back to the modified files.

To be fair, this is a pretty common problem I've hit with container image repositories and registries, if you have separate 'release' and 'auto-build' output locations. There was [a proposal for Helm to make image-registry handling more specific](https://github.com/helm/helm/issues/7754) and allow solving issues like this more streamlined, but it didn't seem to go anywhere.

Reply:
> You can't set variables except .Chart.Version and .Chart.AppVersion when calling helm package. 

(╯°□°)╯︵ ┻━┻

Reply:
There was briefly a feature to do exactly that, but it broke other workflows, e.g., stripped all the comments from the values.yaml, and was backed out.

Reply:
Was thinking about this more, would it make sense to switch to "building" charts?

Basically:

- Copy chart folder from install/ into build/ (which is in .gitignore)
- Build subcharts and put them in the proper folder
- Edit the version and image info
- Build the actual chart

This would also get rid of committed zip file sub charts (which kind of also shows that there are build steps to the charts already, they're just a bit of a hack today.)

Reply:
Yup, that would definitely be an option. And if you're doing that, I'd just copy the current subcharts into the charts/ folder, rather than building them separately and copying the tarballs in, since (as far as I recall) they are not _deployed_ separately anywhere anyway. In effect, Helm's just extracting those those tarballs internally and treating the whole thing as one big directory anyway, so it removes a needless step of "building" the subcharts, when they don't really have an independent life outside the open-match chart.

You'd still run `helm dep up` or `helm dep build` (I forget which one is in the OM workflow) to update external charts, but that won't touch chart _directories_ in the charts/ folder.